### PR TITLE
Issue/3096 masterbar jank reader

### PR DESCRIPTION
**Pull request by nbradbury**
Resolves much of the "masterbar jank" reported in #3096 by optimizing the reader fragment. Changes include:

* Using static booleans rather than EventBus events to determine whether the db has been purged and whether we need to update followed tags/blogs
* Using a separate thread when checking whether it's time to auto-update the current tag
* Optimizing the creation of the post adapter

Last, but certainly not least, this PR changes `WPMainTabAdapter` (the adapter for the main viewpager) from a `FragmentStatePagerAdapter` to a `FragmentPagerAdapter` that keeps fragments in memory by setting the offscreen limit. This way the masterbar fragments aren't destroyed and recreated as the user switches between tabs. I originally didn't do this out of fear that it would consume too much memory, but testing showed that it doesn't significantly increase memory consumption.

### CODE REVIEW

**Comment by maxme** [comment]
I tried to compare memory consumption on `develop` and this branch using `monitor` but I can't even use the app when `monitor` runs (I can switch a tab once in a while).

I can't see much difference between the 2 branches on my G3 (wasn't janky before the patch). But I notice a lot less calls to the GC using this branch.

**Comment by roundhill** [comment]
Trying this out on my janky galaxy ;)

**Comment by roundhill** [comment]
The jank is now gone when returning from an activity! Woot.

I noticed a bug with the Reader:
1. Scroll down in the reader.
2. Switch to a different tab.
3. Tap on the Reader tab, notice that you are scrolled back to the top. I think it should probably just switch back to the tab w/o scrolling.

**Comment by roundhill** [comment]
> I noticed a bug with the Reader:

Same issue with Notifications!

**Comment by nbradbury** [comment]
@roundhill That sounds like [this bug](https://github.com/wordpress-mobile/WordPress-Android/pull/3106), which was introduced in v22 of the support library, then fixed in v23, then re-introduced in v23.0.1 :disappointed: 

We provided a workaround for that (again) [here](https://github.com/wordpress-mobile/WordPress-Android/pull/3162), which has been pulled into this branch. Perhaps you need to do a gradle sync & rebuild?

**Comment by roundhill** [comment]
> Perhaps you need to do a gradle sync & rebuild?

oh, I pulled the wrong branch. My bad! I had pulled `issue/3096-masterbar-jank` and not `issue/3096-masterbar-jank-reader`

**Comment by roundhill** [comment]
After pulling the correct branch, I can see immediate speed improvements in both returning from an activity as well as switching tabs. Nice job @nbradbury!

I've also run it on my craptasticly old Galaxy S running 4.0.3. It is able to load tons of reader posts as well as the full list of notifications w/o any crashes. :+1:

**Comment by nbradbury** [comment]
Thanks for trying this out @roundhill - very relieved to hear this branch did the job without requiring the hacks of that previous branch!

**Comment by roundhill** [comment]
Code looks ok to me, @maxme were you going to reveal as well? Otherwise I say :shipit:

**Comment by nbradbury** [comment]
> Did you find that hitting the db on the main thread was the main cause of the lag?

Hitting the db on the main thread was an issue, but I think the lag was caused by a number of seemingly small things that added up. The [routine](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/datasets/ReaderTagTable.java#L263) which determines whether to update the current tag was a big part of that, but in addition to hitting the database it also did some expensive date formatting. Moving all of that to a separate thread helped a lot.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java [hunk 1]
```diff
@@ -812,6 +814,28 @@ private void updateCurrentTag() {
...
     }
 
+    /*
+     * update the current tag if it's time to do so
+     */
+    private void updateCurrentTagIfTime() {
+        if (!isAdded() || !hasCurrentTag()) {
+            return;
+        }
+        new Thread() {
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostListFragment.java]: Can we add a comment here to explain why we need this (eventually with a ref to that PR).

**Comment by maxme** [comment]
> @maxme were you going to review as well? Otherwise I say

Looks good, just need some comments to tell future-us why it was necessary - so we won't be undoing some part of it.

**Comment by roundhill** [comment]
:shipit:

**Pull request merged**